clarifai/rag/rag.py

```diff
@@ -111,20 +114,112 @@ def setup(cls,
     del user, llm, prompter_model, prompter_model_params
     return cls(workflow=wf)
 
-  # TODO: Implement this.
-  def upload():
-    """Does the following:
+  def upload(self,
+             file_path: str = None,
+             folder_path: str = None,
+             url: str = None,
+             batch_size: int = 128,
+             chunk_size: int = 1024,
+             chunk_overlap: int = 200,
+             **kwargs) -> None:
+    """Uploads documents to the app.
         - Read from a local directory or public url or local filename.
         - Parse the document(s) into chunks.
         - Ingest chunks into the app with metadata.
 
+    Args:
+        file_path str: File path to the document.
+        folder_path str: Folder path to the documents.
+        url str: Public url to the document.
+        batch_size int: Batch size for uploading.
+        chunk_size int: Chunk size for splitting the document.
+        chunk_overlap int: The token overlap of each chunk when splitting.
+        **kwargs: Additional arguments for the SentenceSplitter. Refer https://docs.llamaindex.ai/en/stable/api/llama_index.node_parser.SentenceSplitter.html
+
     Example:
         >>> from clarifai.rag import RAG
         >>> rag_agent = RAG().setup()
-        >>> rag_agent.upload("~/work/docs")
-        >>> rag_agent.upload("~/work/docs/manual.pdf")
+        >>> rag_agent.upload(folder_path = "~/work/docs")
+        >>> rag_agent.upload(file_path = "~/work/docs/manual.pdf")
     """
-    pass
+    #set batch size
+    if batch_size > MAX_UPLOAD_BATCH_SIZE:
+      raise ValueError(f"batch_size cannot be greater than {MAX_UPLOAD_BATCH_SIZE}")
+
+    #check if only one of file_path, folder_path, or url is specified
+    if file_path and (folder_path or url):
+      raise ValueError("Only one of file_path, folder_path, or url can be specified.")
+    if folder_path and (file_path or url):
+      raise ValueError("Only one of file_path, folder_path, or url can be specified.")
+    if url and (file_path or folder_path):
+      raise ValueError("Only one of file_path, folder_path, or url can be specified.")
+
+    #loading documents
+    documents = load_documents(file_path=file_path, folder_path=folder_path, url=url)
+
+    #splitting documents into chunks
+    text_chunks = []
+    metadata = []
+
+    #iterate through documents
+    for doc in documents:
+      cur_text_chunks = split_document(
+          text=doc.text, chunk_size=chunk_size, chunk_overlap=chunk_overlap, **kwargs)
+      text_chunks.extend(cur_text_chunks)
+      metadata.extend([doc.metadata for _ in range(len(cur_text_chunks))])
+      #if batch size is reached, upload the batch
+      if len(text_chunks) > batch_size:
+        for idx in range(0, len(text_chunks), batch_size):
+          if idx + batch_size > len(text_chunks):
+            continue
```

Comment:
Just wondering what this check is for?

Reply:
This is to upload to the app as soon as batch size is full. So checking if the number of chunks from current iteration is less than the `batch_size`, If it is lesser, then it will continue to chunk the documents.